src/appellate/appellate.js

```diff
@@ -200,3 +206,179 @@ AppellateDelegate.prototype.handleAttachmentPage = async function () {
+AppellateDelegate.prototype.handleSingleDocumentPageView = async function () {
+  overwriteFormSubmitMethod();
+
+  this.pacer_case_id = await APPELLATE.getCaseId(this.tabId, this.queryParameters, this.docId);
+  await saveCaseIdinTabStorage({ tabId: this.tabId }, this.pacer_case_id);
+
+  // When we receive the message from the above submit method, submit the form
+  // via XHR so we can get the document before the browser does.
+  window.addEventListener('message', this.onDocumentViewSubmit.bind(this), false);
+
+  let callback = $.proxy(function (api_results) {
+    console.info(`RECAP: Got results from API. Running callback on API results to ` + `insert link`);
+    let result = api_results.results.filter((obj) => obj.pacer_doc_id == this.docId, this)[0];
+
+    if (!result) {
+      return;
+    }
+
+    let href = `https://www.courtlistener.com/${result.filepath_local}`;
+    // Insert a RECAP download link at the bottom of the form.
+    $('<div class="recap-banner"/>')
+      .append(
+        $('<a/>', {
+          title: 'Document is available for free in the RECAP Archive.',
+          href: href,
+        })
+          .append($('<img/>', { src: chrome.extension.getURL('assets/images/icon-16.png') }))
+          .append(' Get this document for free from the RECAP Archive.')
+      )
+      .appendTo($('body'));
+  }, this);
+
+  this.recap.getAvailabilityForDocuments([this.docId], this.court, callback);
+};
+
+AppellateDelegate.prototype.onDocumentViewSubmit = function (event) {
+  // Security check to ensure message is from a PACER website.
+  if (!PACER.getCourtFromUrl(event.origin)) {
+    console.warn(
+      'Received message from non PACER origin. This should only ' +
+        'happen when the extension is being abused by a bad actor.'
+    );
+    return;
+  }
+
+  let previousPageHtml = copyPDFDocumentPage();
+  let form = document.getElementById(event.data.id);
+
+  let title = document.querySelectorAll('strong')[1].innerHTML;
+  let dataFromTitle = APPELLATE.parsePdfDataFromTitle(title);
+  if (!dataFromTitle) {
+    form.submit();
+    return;
+  }
+  $('body').css('cursor', 'wait');
+  let query_string = new URLSearchParams(new FormData(form)).toString();
+  httpRequest(
+    form.action,
+    query_string,
+    'application/x-www-form-urlencoded',
+    function (type, ab, xhr) {
+      console.info(`RECAP: Successfully submitted RECAP "View" button form: ${xhr.statusText}`);
+
+      const blob = new Blob([new Uint8Array(ab)], { type: type });
+      // If we got a PDF, we wrap it in a simple HTML page.  This lets us treat
+      // both cases uniformly: either way we have an HTML page with an <iframe>
+      // in it, which is handled by showPdfPage.
+      if (type === 'application/pdf') {
+        // canb and ca9 return PDFs and trigger this code path.
+        let html = APPELLATE.getFullPageIframe(URL.createObjectURL(blob));
+        this.showPdfPage(
```

Comment:
A lot of this and the stuff below is repeated, no? Why not factor it out better? Is it too different?

Reply:
The callback functions for this method are the same after we merge the two approaches to create the full-page Iframe, so we can refactor this code into a helper function to make the main method shorter and more readable.